public/apps/configuration/utils/request-utils.ts

```diff
@@ -13,29 +13,54 @@
  *   permissions and limitations under the License.
  */
 
-import { HttpStart, HttpHandler } from 'opensearch-dashboards/public';
+import { HttpStart, HttpHandler, HttpFetchQuery } from 'opensearch-dashboards/public';
 
-export async function request<T>(requestFunc: HttpHandler, url: string, body?: object): Promise<T> {
+interface RequestType {
```

Comment:
IMO I think its clearer when all of the params are listed out instead of having these 2 interfaces. 

RequestParams makes more sense than RequestType as a request would have an HTTP Verb, url, body and query params. The RequestType interface does not make sense.

Reply:
The problem with the old implementation is that the request function takes in a possibly undefined body, then on top of that we would need to add a possibly undefined query param. Personally I think it is bad practice to have a function signature take in two possibly undefined things without forcing users of that function to explicitly list out that they are passing/not passing in a specific variable. Does that make sense/maybe there is another way of doing this.

Reply:
RequestType as an interface doesn't make sense to me, but RequestParams does. I can see how RequestType is convenient in the context of this method, but an HttpStart would not be enough to specify a webrequest. From its usage below it looks like you can call http.get for a GET request or http.post for a POST request. 

RequestParams makes more sense as an interface since you can fully define a request with those params, RequestType is for convenience.

i.e.

Verb: GET
url: `/api/internalusers`
body: None
query: `dataSource=dataSource1`

Reply:
its more of a wrapper than anything else to allow strict mapping of values, particularly for those which can be null/undefined. i.e. if 3rd arg is undefined, but 4th is defined, this interface will ensure that these values are correctly mapped to intended variables.

Reply:
I think part of the confusion is coming from the very similarly named/similar parameters. One is expecting a `httpHandler`, and one is expecting a `httpStart`. The one that is expecting a `httpStart` is simply extracting a function off of it and passing it in as the `httpHandler` of the other. The way the code is written now, I think we need both - but I think later on we can probably remove all `httpGet`, and simply replace it with a generic `request`, with `http.get` passed in, but that seems to be a larger refactor effort later on. I can create an issue to capture this - does that help @cwperks ? But yea +1 to @DarshitChanpura in that this is simply a wrapper type and the naming seems off and possibly creating confusion.

Reply:
Note; I think this is critical to land before merging this PR.  @cwperks what do you think?

Reply:
@peternied the change here is relatively simple. To allow httpGet, httpPost, etc methods to accept one more param `dataSourceQuery` which is a Json object. The issue with simply adding a dataSourceQuery as an argument is that not all previous arguments in each caller are expected to be present, some may not have any value passed. Thus, in order for these http calls to correctly pick up corresponding values (i.e., url, body, additionalHeaders, etc), we extracted arguments into an object annotation. This allowed us the flexibility to accept arguments with correctly mapped values. From what I understand end user is not concerned with this change and will not have visibility into this. This was the simplest way to handle the arguments passed to http functions that are used internally. @derek-ho

Reply:
I would expect the interface to be defined on the Request level with the parameters being included as part of the struct. For example

```
interface Request {
  httpFunc: string;
  url: string;
  body?: object;
  query?: HttpFetchQuery;
}
```

Then define a util that processes the Request.httpFunc to resolve into the HttpHandler you are using already. This avoid duplicating the interfaces and the "nesting" that this makes. 

Perhaps there is some reason this is not feasible but I would expect this to work based on the review?

Reply:
I addressed the main concern hereby refactoring through a helper class that exposes the method, and removing the public implementation of it. That forced all usages to use the new class and pass in the datasourceId for more confidence on this change. I think the concern with the naming of the new interface is a minor thing and should not prevent this change from going in. @cwperks would you be ok if I create an issue to do a fast follow up on this PR?

Reply:
Resolving this now